### PR TITLE
Refactor scan job service make it easy to add new scan type

### PR DESCRIPTION
**Pull request by stonezdj**
Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
Fixes #(issue)

Please indicate you've done the following:
- [ ] Well Written Title and Summary of the PR
- [ ] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [ ] Accepted the DCO. Commits without the DCO will delay acceptance.
- [ ] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: src/pkg/scan/vulnerability/vul.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/goharbor/harbor/src/pkg/permission/types"
+	"github.com/goharbor/harbor/src/pkg/robot/model"
+	scanJob "github.com/goharbor/harbor/src/pkg/scan"
+	"github.com/goharbor/harbor/src/pkg/scan/dao/scan"
+	"github.com/goharbor/harbor/src/pkg/scan/postprocessors"
+	v1 "github.com/goharbor/harbor/src/pkg/scan/rest/v1"
+)
+
+func init() {
+	scanJob.RegisterScanHanlder(v1.ScanTypeVulnerability, &ScanHandler{})
```
Comment (zyyw) [inline src/pkg/scan/vulnerability/vul.go:31]: As we see there is only `v1.ScanTypeVulnerability` registered in this PR, I assume that this PR has nothing to do with SBOM but refactoring the current scan for vulnerabilities logic merely.

File: src/controller/scan/options.go [hunk 2]
```diff
@@ -44,3 +55,11 @@ func WithTag(tag string) Option {
 		return nil
 	}
 }
+
+// WithScanType set the scanType
+func WithScanType(scanType string) Option {
+	return func(options *Options) error {
```
Comment (zyyw) [inline src/controller/scan/options.go:61]: do we need to validate the param `scanType` before returning an `Option`?

File: src/pkg/scan/handler.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/goharbor/harbor/src/pkg/robot/model"
+	"github.com/goharbor/harbor/src/pkg/scan/dao/scan"
+	v1 "github.com/goharbor/harbor/src/pkg/scan/rest/v1"
+)
+
+var handlerRegistry = map[string]Handler{}
+
+// RegisterScanHanlder register scanner handler
+func RegisterScanHanlder(requestType string, handler Handler) {
+	handlerRegistry[requestType] = handler
```
Comment (zyyw) [inline src/pkg/scan/handler.go:31]: validate `requestType` before registering the scanHandler for it?

File: src/controller/scan/options.go [hunk 4]
```diff
@@ -44,3 +55,11 @@ func WithTag(tag string) Option {
 		return nil
 	}
 }
+
+// WithScanType set the scanType
+func WithScanType(scanType string) Option {
```
Comment (zyyw) [inline src/controller/scan/options.go:60]: `WithScanType` is only defined here, but not invoked anywhere?

**Review by wy65701436 — APPROVED** [review]
lgtm

File: src/pkg/scan/vulnerability/vul.go (hunk 1 above)
Comment (stonezdj) [reply src/pkg/scan/vulnerability/vul.go:31]: Correct

File: src/pkg/scan/handler.go (hunk 3 above)
Comment (stonezdj) [reply src/pkg/scan/handler.go:31]: request type is const string

File: src/controller/scan/options.go (hunk 2 above)
Comment (stonezdj) [reply src/controller/scan/options.go:61]: const value

**Review by zyyw — APPROVED** [review]

**Pull request merged**